### ISSUE TITLE
Logprep applies some rules multiple times if they match

### ISSUE THREAD

**Issue by ppcad** [fkie-cad/Logprep#166]
Logprep applies some rules multiple times if they matches.
It's one time for each matching disjunction a rule filter contains.

**Expected behavior**
Logprep applies a rule only once if it matches.

**Current behavior**
Logprep applies a rule multiple times for each matching disjunction it's rule filter contains.

**Steps to reproduce**

1. Create a rule with a fiter like "term_1 OR term_2", where both terms are true for some event X. This is most easily tested with a processor that does not allow duplication (i.e. the domain resolver).
2. Run Logprpe and send event X.
3. See the duplication warning in the console.

**Environment**

Logprep version:
Python version: >=3.6

**Possible solution**
Adapt the rule tree to not match a rule multiple times or remove duplicates after the matching process if possible.

**Comment by ekneg54** [comment]
in every processor rule we define, how equality is set for the rule. I do not think, that this happens to all processors, but we have to check that.
It could be possible, that we forget about this behavior by defining equality in rules.

**Comment by ekneg54** [comment]
especially  for the domain_resolver we have no rule tests and the equality of two domain_resolver rules is never testet. :-/

**Comment by ekneg54** [comment]
we could dissect a rule with a logical OR expression in filter statement into two rules where the dissected rules have one part of the filter expression as filter and the same other fields. so these rules are not processed multiple times for one event?

**Comment by ekneg54** [comment]
is it as simple as I think, that we should return a `set` here?:

https://github.com/fkie-cad/Logprep/blob/6ddbe4875998ed00e3065e3aaa083befebd3db6a/logprep/framework/rule_tree/rule_tree.py#L178

**Comment by ppcad** [comment]
Thanks, but I have already fixed it. I just need to do some more testing.
There was no check for duplicates in the rule tree and the equality check was not done correctly for some rules types.
I removed some duplication that happened already at the time of rule tree creation (in add_rule) and then ensured that there is no duplication in get_matching_rules as well.

One thing I'm still questioning is how to handle rules with different definitions but equal results. There are some rules that have different rule definitions, but end up having equal effects. Like for the generic adder with "add_from_file". There it is possible that the results from the file are the same as if it was added via a different configuration. I decided to make rules equal if the results of their configuration is equal.
